### ISSUE TITLE
Package is broken with node 14 : fs.watch with recursive is throwing on incompatible os

### ISSUE THREAD

**Issue by daeschlerdavid** [whitecolor/ts-node-dev#143]
Hello,

I just upgraded node to v14 on my fedora 31.
It seems this package is now broken due to this commit : https://github.com/nodejs/node/commit/67e067eb06

Also check updated doc : [fs.watch doc](https://nodejs.org/api/fs.html#fs_fs_watch_filename_options_listener)

> The recursive option is only supported on macOS and Windows. An ERR_FEATURE_UNAVAILABLE_ON_PLATFORM exception will be thrown when the option is used on a platform that does not support it.

Output error example :
```
# npx ts-node-dev src/main.ts
Using ts-node version 8.4.1, typescript version 3.8.3
TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
    at Object.watch (fs.js:1441:11)
    at add (/root/backend/node_modules/filewatcher/index.js:74:34)
    at /root/backend/node_modules/filewatcher/index.js:93:5
    at FSReqCallback.oncomplete (fs.js:176:5)
```

Using recursive: true now throws an Error that is not handled.
https://github.com/whitecolor/ts-node-dev/blob/c0df64c43455d2f8ef144eb9defcd05b6fc52486/lib/index.js#L45

**Comment by SrBrahma** [comment]
Same issue here. Thanks for pointing out that the problem is caused by the Node v14. Would never figure out.

**Comment by Leko** [comment]
@whitecolor I think we can fix this issue by migrating to chokidar instead of fs.watch. Does it make sense?

**Comment by desmap** [comment]
facing the same problem

**Comment by Kodbollen** [comment]
I was facing the same problem and replaced the `recursive:true` flag in the _watcher_ variable in my _node_modules/ts-node-dev/lib/index.js_ with  `recursive: process.platform !== 'linux'. The variable will then be:
```javascript
var watcher = filewatcher({
    forcePolling: opts.poll,
    interval: parseInt(opts.interval),
    debounce: parseInt(opts.debounce),
    recursive: process.platform !== 'linux'
  })
```
I am very new to Type- and Javascript and don't know the intricacies of the ts-node-dev project nor fs. Therefore, I am not sure this is a proper fix, but it seems to be working just fine.

**Comment by sarpik** [comment]
Bump - just updated to node v14 and got the same issue.

**Comment by McSneaky** [comment]
Issue is in underlying [`filewatcher`](https://www.npmjs.com/package/filewatcher) issue is already raised in there and [PR made](https://github.com/fgnass/filewatcher/issues/5)

Thnx for @onlurking for making fix for it

But it seems like filewatcher has not been updated for a while tho, so not sure if it will get merged in there :disappointed:

**Comment by marcoreni** [comment]
Maybe this project should migrate to chokidar:

```
Node.js fs.watch:
* Doesn't report filenames on MacOS.
* Doesn't report events at all when using editors like Sublime on MacOS.
* Often reports events twice.
* Emits most changes as rename.
* Does not provide an easy way to recursively watch file trees.

Node.js fs.watchFile:
* Almost as bad at event handling.
* Also does not provide any recursive watching.
* Results in high CPU utilization.

Chokidar resolves these problems.
```

**Comment by whitecolor** [comment]
Will check everything and release it with chokidar soon.

**Comment by kachkaev** [comment]
@whitecolor if you are planning to release a new `×.−.−` or `−.×.−` version, will be happy to also accept a PR that adds #136? I can do it if there is a probability of acceptance 🙂

**Comment by thangbn** [comment]
looking forward to the release

**Comment by mmvsk** [comment]
Hey, while waiting for a fix in [filewatcher](https://github.com/fgnass/filewatcher/issues/5) or here, I just made a [quick & dirty fix](https://github.com/mmvsk/ts-node-dev/commit/d7e9cfbad414bcc5fc9ddb5cf75a74496a9e4549) by disabling the `recursive` option on Linux and Node 14, to be able to use `ts-node-dev` right now in this environment.

To select this specific commit: `npm add -D mmvsk/ts-node-dev#d7e9cfbad414bcc5fc9ddb5cf75a74496a9e4549`

**Comment by ssomlk** [comment]
I am getting this error when using node:alpine when using Docker.
Suppose I need to downgrade the Node version how can I do it in the Docker file.

SInce I am new to docker would appreciate if someone can assist

**Comment by marcoreni** [comment]
@ssomlk you should `node:13-alpine` if you want to use a different version.

Here you can find a list of all the released tags: https://hub.docker.com/_/node/
I think that Docker has a Get Started guide that may help you with these basic concepts.

**Comment by ssomlk** [comment]
Thanks alot and apreciate ur help @marcoreni

**Comment by mythio** [comment]
Run using `ts-node-dev --poll src/index.ts`

**Comment by mpcen** [comment]
using node v13.14.0 until the fix is merged and works fine

**Comment by DonIsaac** [comment]
@mmvsk That fix is pretty nifty, have you submitted it as a pull request yet?

**Comment by mmvsk** [comment]
> @mmvsk That fix is pretty nifty, have you submitted it as a pull request yet?

I just did in PR #149. I'm using `ts-node-dev` with this fix and everything works just fine (including in deep subdirectories).

**Comment by loreanvictor** [comment]
@mmvsk doesn't your PR simply disable recursive watching on linux / node 14? or would it work on linux machines with recursive watching as well (and I am mis-understanding something)?

P.S. perhaps using something like [this](https://www.npmjs.com/package/node-watch) (or another alternative `fs.watch` wrapper, I don't know which of these packages actually solves this issue and is performant/reliable enough) would alleviate the issue?

**Comment by mmvsk** [comment]
> @mmvsk doesn't your PR simply disable recursive watching on linux / node 14? or would it work on linux machines with recursive watching as well (and I am mis-understanding something)?
> 
> P.S. perhaps using something like [this](https://www.npmjs.com/package/node-watch) (or another alternative `fs.watch` wrapper, I don't know which of these packages actually solves this issue and is performant/reliable enough) would alleviate the issue?

Yes, it's precisely what it does. But actually the recursive option never worked on Linux: before Node 14 `fs-watch` was silently failing, and now it raise an error. So this PR is basically reverting `ts-node-dev` to its original behavior.

It "works" on Linux, in the same way it was working before Node 14, thanks to the dependency chain built by the imports. So it's possible to have some missed detections if a file is modified in a deep directory without being linked by imports - in the exact same way as in Node 13 and before.

And I agree with you: ideally, `fs-watch` should be replaced and [`node-watch`](https://github.com/yuanchuan/node-watch) seems like a good candidate for that, better than `chokidar` that was proposed earlier in this thread.

**tl;dr**

- PR #149 fix this actual issue, by restoring the same behavior as before Node 14
- As an improvement, `fs-watch` should be replaced by a better alternative (such as `node-watch`)

**Comment by whitecolor** [comment]
Replaced with `chokidar` in the latest published.

**Comment by FredyC** [comment]
Thank you for the fix. Had the same issue on Windows and WSL2 and the latest `pre.49` works nicely.
